lua/starfall/libs_cl/bass.lua

```diff
@@ -1,20 +1,107 @@
 local checkluatype = SF.CheckLuaType
 local registerprivilege = SF.Permissions.registerPrivilege
+local math_sqrt = math.sqrt
 
 -- Register privileges
 registerprivilege("bass.loadFile", "Play local sound files with `bass`.", "Allows users to create sound channels by file path.", { client = {} })
 registerprivilege("bass.loadURL", "Play remote sound files with `bass`.", "Allows users to create sound channels by URL.", { client = {}, urlwhitelist = {} })
 registerprivilege("bass.play2D", "Play sounds in global game context with `bass`.", "Allows users to create sound channels which play in global game context (without `3d` flag).", { client = { default = 1 } })
 
 local plyCount = SF.LimitObject("bass", "bass sounds", 20, "The number of sounds allowed to be playing via Starfall client at once")
+local soundDatas = {} -- { [IGModAudioChannel] = { ... }, ... } -- Contains extra data for each starfall sound.
+local simpleFadeSounds = {} -- { [1] = IGModAudioChannel, ... } -- A list of sounds that need to be manually controlled through 'simple fading'.
 
 SF.ResourceCounters.Bass = {icon = "icon16/sound_add.png", count = function(ply) return plyCount:get(ply).val end}
 
-local function deleteSound(ply, sound)
-	if sound:IsValid() then sound:Stop() end
+
+-- Calculates whether or not a 3D sound is out of earshot, and its fade multiplier.
+local function getSimpleFading(snd)
+	local pos = snd:GetPos()
+	local earPos = EyePos()
+	local fadeMin = snd:Get3DFadeDistance()
+	local fadeMax = soundDatas[snd].fadeMax
+	local distSqr = pos:DistToSqr(earPos)
+
+	if distSqr <= fadeMin * fadeMin then return false, 1 end
+	if distSqr >= fadeMax * fadeMax then return true, 0 end
+
+	local dist = math_sqrt(distSqr)
+	local farnessFrac = (dist - fadeMin) / (fadeMax - fadeMin)
+
+	return false, (1 - farnessFrac) ^ 2 -- Sounds falls off with dist^2. Unfortunately, we still have to do the square root inbetween.
+end
+
+local function applySimpleFading(snd)
+	local sndData = soundDatas[snd]
+	local outOfEarshot, fadeMult = getSimpleFading(snd)
+
+	if sndData.fadeMult ~= fadeMult then
+		sndData.fadeMult = fadeMult
+		sndData.outOfEarshot = outOfEarshot
+
+		snd:SetVolume(sndData.targetVolume * fadeMult)
+	end
+end
+
+local function addSoundToSimpleFade(snd)
+	local sndData = soundDatas[snd]
+	if sndData.simpleFadeEnabled then return end
+
+	sndData.simpleFadeEnabled = true
+	table.insert(simpleFadeSounds, snd)
+	applySimpleFading(snd)
+
+	if #simpleFadeSounds ~= 1 then return end -- Hook is already running.
+
+	hook.Add("Think", "SF_Bass_SimpleFade", function()
+		for _, s in ipairs(simpleFadeSounds) do
+			if s:IsValid() and s:GetState() == GMOD_CHANNEL_PLAYING then
+				applySimpleFading(s)
+			end
```

Comment:
It really should be `for snd, sndData in pairs(simpleFadeSounds) do` and `applySimpleFading(snd, sndData)` so that applySimpleFading doesn't need to do a lookup for sndData.

Reply:
This uses a list since `ipairs()` is faster than `pairs()`, and we're doing it every tick.

Honestly, the `:IsValid()` check could be removed, nothing should be able to invalidate the sound without calling `deleteSound()`.
A quick test affirms that `snd_restart`, `stopsound`, and `record fix;stop` all preserve sound objects.